File: python/0977_squares_of_a_sorted_array.py
# https://leetcode.com/problems/squares-of-a-sorted-array/
# O(n) time complexity
# O(n) space complexity
def sorted_squares(nums):
    squares = []

    pos_neg_boundary = 0
    while pos_neg_boundary < len(nums) and nums[pos_neg_boundary] < 0:
        pos_neg_boundary += 1
    left = pos_neg_boundary - 1
    right = pos_neg_boundary

    while left >= 0 and right < len(nums):
        if abs(nums[left]) < abs(nums[right]):
            squares.append(nums[left] ** 2)
            left -= 1
        else:
            squares.append(nums[right] ** 2)
            right += 1
    while left >= 0:
        squares.append(nums[left] ** 2)
        left -= 1
    while right < len(nums):
        squares.append(nums[right] ** 2)
        right += 1

    return squares

File: leetcode/java/0977_squares_of_a_sorted_array.java
// https://leetcode.com/problems/squares-of-a-sorted-array/
// O(n) time complexity
// O(n) space complexity to store output
class Solution_0977 {
  public int[] sortedSquares(int[] nums) {
    int[] sortedSquareArray = new int[nums.length];
    int left = 0, right = nums.length - 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      int largerMagnitude;
      if (Math.abs(nums[left]) < Math.abs(nums[right])) {
        largerMagnitude = nums[right];
        right--;
      } else {
        largerMagnitude = nums[left];
        left++;
      }
      sortedSquareArray[i] = largerMagnitude * largerMagnitude;
    }
    return sortedSquareArray;
  }
}